src/ocrd/processor/base.py

```diff
@@ -188,20 +193,122 @@ def verify(self):
+                for input_file_tuple in self.zip_input_files(on_error='abort'):
+                    # FIXME: add error handling by catching exceptions in various ways (#579)
+                    # for example:
+                    # - ResourceNotFoundError → use ResourceManager to download (once), then retry
+                    # - transient (I/O or OOM) error → maybe sleep, retry
+                    # - persistent (data) error → skip / dummy / raise
+                    input_files = [None] * len(input_file_tuple)
+                    for i, input_file in enumerate(input_file_tuple):
+                        if i == 0:
+                            log.info("processing page %s", input_file.pageId)
+                        elif input_file is None:
+                            # file/page not found in this file grp
+                            continue
+                        input_files[i] = input_file
+                        if not self.download:
+                            continue
+                        try:
+                            input_files[i] = self.workspace.download_file(input_file)
+                        except ValueError as e:
+                            log.error(repr(e))
+                            log.warning("skipping file %s for page %s", input_file, input_file.pageId)
+                    self.process_page_file(*input_files)
+            except NotImplementedError:
+                # fall back to deprecated method
+                self.process()
 
-    def add_metadata(self, pcgts):
+    def process_page_file(self, *input_files) -> None:
+        """
+        Process the given ``input_files`` of the :py:attr:`workspace`,
+        representing one physical page (passed as one opened
+        :py:class:`~ocrd_models.OcrdFile` per input fileGrp)
+        under the given :py:attr:`parameter`, and make sure the
+        results get added accordingly.
+
+        (This uses process_page_pcgts, but can be overridden by subclasses
+        to handle cases like multiple fileGrps, non-PAGE input etc.)
+        """
+        log = getLogger('ocrd.processor.base')
+        input_pcgts = [None] * len(input_files)
+        for i, input_file in enumerate(input_files):
+            # FIXME: what about non-PAGE input like image or JSON ???
+            log.debug("parsing file %s for page %s", input_file.ID, input_file.pageId)
+            try:
+                input_pcgts[i] = page_from_file(input_file)
+            except ValueError as e:
+                log.info("non-PAGE input for page %s: %s", input_file.pageId, e)
+        output_pcgts = self.process_page_pcgts(*input_pcgts)
+        output_file_id = make_file_id(input_files[0], self.output_file_grp)
+        output_pcgts.set_pcGtsId(output_file_id)
+        self.add_metadata(output_pcgts)
+        # FIXME: what about save_image_file in process_page ???
+        # FIXME: what about non-PAGE output like JSON ???
```

Comment:
Could you name some processors that produce such an output so I could check the implementation?

Reply:
- ocrd-segment-evaluate
- ocrd-cor-asv-ann-evaluate
- ocrd-segment-extract-*
- ocrd-fileformat-transform (depending on the output type)
- ocrd-pagetopdf